### PR TITLE
New package: Sarus

### PR DESCRIPTION
**Pull request by teonnik**
Sarus is an OCI-compliant container engine for HPC systems.

The post-install step requires root privileges. The package issues a warning at the end of spack's installation with the necessary steps to complete the configuration (similar to var/spack/repos/builtin/packages/singularityce/package.py)

tini is a tiny init for containers. It is a dependency of Sarus.

As an alternative to the post-installation message, I was trying to call sudo from within Sarus's package.py on the script mentioned in the post-installation message so as to avoid the chance of a user missing the message. The command ran fine but it didn't accept valid password entries. That is perhaps because spack is processing the input before is handed over to sudo?

### CODE REVIEW

File: var/spack/repos/builtin/packages/rapidjson/package.py [hunk 1]
```diff
@@ -12,14 +12,17 @@ class Rapidjson(CMakePackage):
     homepage = "https://rapidjson.org"
     url      = "https://github.com/Tencent/rapidjson/archive/v1.1.0.tar.gz"
 
+    version('00dbcf2',
+            git = 'https://github.com/Tencent/rapidjson.git',
+            commit = '00dbcf2')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rapidjson/package.py]:
```suggestion
    version('1.2.0-2021-08-13', git='https://github.com/Tencent/rapidjson.git', commit='00dbcf2c6e03c47d6c399338b6de060c71356464')
```

This is a bit improvised... but at least this way the versions are properly ordered and more releases between 2016 and now can be added...

File: var/spack/repos/builtin/packages/sarus/package.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    version("master", branch="master")
+    version("1.4.0", tag="1.4.0")
+    version("1.3.3", tag="1.3.3")
+    version("1.3.2", tag="1.3.2")
+    version("1.3.1", tag="1.3.1")
+    version("1.3.0", tag="1.3.0")
+    version("1.2.0", tag="1.2.0")
+    version("1.1.0", tag="1.1.0")
+    version("1.0.1", tag="1.0.1")
+    version("1.0.0", tag="1.0.0")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/sarus/package.py]: Any reason not to use release tarballs here? Then you get checksums and better source caching.

File: var/spack/repos/builtin/packages/sarus/package.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        "to enable connections inside containers",
+    )
+
+    depends_on("wget", type="build")
+    depends_on("expat", type="build")
+    depends_on("squashfs", type=("build", "run"))
+    depends_on("boost@1.65.0: cxxstd=11")
+    depends_on("cpprestsdk@2.10.0:")
+    depends_on("libarchive@3.4.1:")
+    depends_on("rapidjson@00dbcf2", type="build")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/sarus/package.py]:
```suggestion
    depends_on("rapidjson@1.2.0-2021-08-13", type="build")
```

File: var/spack/repos/builtin/packages/sarus/package.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    version("1.0.0", tag="1.0.0")
+
+    variant(
+        "ssh",
+        default=False,
+        description="Build and install the SSH hook and custom SSH software "
+        "to enable connections inside containers",
+    )
+
+    depends_on("wget", type="build")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/sarus/package.py]: Why wget actually?

File: var/spack/repos/builtin/packages/sarus/package.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    depends_on("libarchive@3.4.1:")
+    depends_on("rapidjson@00dbcf2", type="build")
+    depends_on("runc")
+    depends_on("tini")
+
+    # autoconf is required to build Dropbear for the SSH hook
+    depends_on("autoconf", type="build")
+
+    # Python 3 is used to run integration tests
+    depends_on("python@3:", type="run", when="@develop")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/sarus/package.py]:
```suggestion
    depends_on("python@3:", type="test", when="@develop")
```

File: var/spack/repos/builtin/packages/sarus/package.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            mkdirp(prefix.var.OCIBundleDir)
+
+    @run_after("install")
+    def build_perms_script(self):
+        script_sh = join_path(self.spec.prefix, "configure_installation.sh")
+        tty.warn(
+            """
+                To complete Sarus's configuration:
+
+                1. Make sure `tini`, `tini-static` and `squashfs` binaries are in PATH, for example do `spack load tini squashfs`.
```
Comment (haampie) [inline var/spack/repos/builtin/packages/sarus/package.py]:
```suggestion
                1. Make sure `sarus` and its dependencies are loaded, for example do `spack load sarus`.
```

**Review by haampie — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/sarus/package.py (hunk 4 above)
Comment (teonnik) [reply var/spack/repos/builtin/packages/sarus/package.py]: It's not needed, good catch! wget was necessary when tini and runc were installed from within the sarus package. However they are now separate packages in spack and are installed as proper dependencies.

**Review by haampie — APPROVED** [review]

**Pull request merged**